File: linear-and-multivariable-algebra/programs/gram_schmidt.py
import numpy as np

def gram_schmidt(vectors):
    basis = []
    for v in vectors:
        w = v - sum(np.dot(v, b) * b for b in basis)
        if (np.linalg.norm(w) > 1e-10):
            basis.append(w / np.linalg.norm(w))
    return np.array(basis)

test = [np.array([1, 1]), np.array([1, 2])]
print(gram_schmidt(test))

File: linear-and-multivariable-algebra/programs/gram_schmidt.cpp
#include "linalg.hpp"

int main() {
    std::vector<linalg::Vector> basis = {{1, 1, 0}, {1, 0, 1}, {0, 1, 1}};
    std::vector<linalg::Vector> orthogonal;
    
    for(const auto& v : basis) {
        linalg::Vector u = v;
        for(const auto& q : orthogonal) {
            double scalar = linalg::dot(v, q) / linalg::dot(q, q);
            for(size_t i=0; i<u.size(); ++i) u[i] -= scalar * q[i];
        }
        orthogonal.push_back(u);
    }
    
    for(size_t i=0; i<orthogonal.size(); ++i) {
        linalg::print(orthogonal[i], "u" + std::to_string(i+1));
    }
    
    return 0;
}
